frontend/test/pytest/test_adjoint.py

```diff
@@ -0,0 +1,276 @@
+        return qml.state()
+
+    @qml.qnode(qml.device("default.qubit", wires=2))
+    def PL_workflow():
+        qml.RX(pnp.pi / 2, wires=0)
+        PL_adjoint(partial(func, A=PL_adjoint, I=0))()
+        qml.RZ(pnp.pi / 2, wires=0)
+        return qml.state()
+
+    actual = C_workflow()
+    desired = PL_workflow()
+    assert_allclose(actual, desired)
+
+
+def test_adjoint_qubitunitary():
+    """Ensures that catalyst.adjoint supports QubitUnitary oprtations."""
+
+    def func():
+        qml.QubitUnitary(
+            jnp.array(
+                [
+                    [0.99500417 - 0.09983342j, 0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
+                    [0.0 + 0.0j, 0.99500417 + 0.09983342j, 0.0 + 0.0j, 0.0 + 0.0j],
+                    [0.0 + 0.0j, 0.0 + 0.0j, 0.99500417 + 0.09983342j, 0.0 + 0.0j],
+                    [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j, 0.99500417 - 0.09983342j],
+                ]
+            ),
+            wires=[0, 1],
+        )
+
+    @qjit()
+    @qml.qnode(qml.device("lightning.qubit", wires=2))
+    def C_workflow():
+        C_adjoint(func)()
+        return qml.state()
+
+    @qml.qnode(qml.device("default.qubit", wires=2))
+    def PL_workflow():
+        PL_adjoint(func)()
+        return qml.state()
+
+    actual = C_workflow()
+    desired = PL_workflow()
+    assert_allclose(actual, desired)
+
+
+def test_adjoint_multirz():
+    """Ensures that catalyst.adjoint supports MultiRZ oprtations."""
+
+    def func():
+        qml.PauliX(0)
+        qml.MultiRZ(theta=pnp.pi / 2, wires=[0, 1])
+
+    @qjit()
+    @qml.qnode(qml.device("lightning.qubit", wires=2))
+    def C_workflow():
+        C_adjoint(func)()
+        return qml.state()
+
+    @qml.qnode(qml.device("default.qubit", wires=2))
+    def PL_workflow():
+        PL_adjoint(func)()
+        return qml.state()
+
+    actual = C_workflow()
+    desired = PL_workflow()
+    assert_allclose(actual, desired)
+
+
+def test_adjoint_no_measurements():
+    """Checks that catalyst.adjoint rejects functions containing quantum measurements."""
+
+    def func():
+        qml.RX(pnp.pi / 2, wires=0)
+        qml.sample()
+
+    with pytest.raises(ValueError, match="no measurements"):
+
+        @qjit()
+        @qml.qnode(qml.device("lightning.qubit", wires=2))
+        def C_workflow():
+            C_adjoint(func)()
+            return qml.state()
+
+        C_workflow()
+
+
+def test_adjoint_invalid_argument():
+    """Checks that catalyst.adjoint rejects non-quantum program arguments."""
+    with pytest.raises(ValueError, match="Expected a callable"):
+
+        @qjit()
+        @qml.qnode(qml.device("lightning.qubit", wires=2))
+        def C_workflow():
+            C_adjoint(33)()
+            return qml.state()
+
+        C_workflow()
+
+
+def test_adjoint_classical_loop():
+    """Checks that catalyst.adjoint supports purely-classical Control-flows."""
+
+    def func(w=0):
+        @for_loop(0, 2, 1)
+        def loop(i, s):
```

Comment:
The following should work now:
```suggestion
        def loop(s):
```

Reply:
I don't think so. But to avoid it, you should use instead:

```suggestion
        def loop(_i, s):
```